transaction/std/src/memo_builder/gift_code_cancellation_memo_builder.rs

```diff
@@ -0,0 +1,215 @@
+/// when the gift code was funded.
+#[derive(Clone, Debug)]
+pub struct GiftCodeCancellationMemoBuilder {
+    // Index of the gift code TxOut that was originally funded
+    gift_code_tx_out_global_index: Option<u64>,
+    // Whether or not to enable change memos
+    gift_code_change_memo_enabled: bool,
+    // Whether we've already written the change memo
+    wrote_change_memo: bool,
+}
+
+// Create an empty GiftCodeCancellationMemoBuilder
+impl Default for GiftCodeCancellationMemoBuilder {
+    fn default() -> Self {
+        Self {
+            gift_code_tx_out_global_index: None,
+            gift_code_change_memo_enabled: true,
+            wrote_change_memo: false,
+        }
+    }
+}
+
+impl GiftCodeCancellationMemoBuilder {
+    /// Set the index of the gift code TxOut that was cancelled
+    pub fn set_gift_code_tx_out_index(&mut self, tx_out_global_index: u64) {
+        self.gift_code_tx_out_global_index = Some(tx_out_global_index)
+    }
+    /// Clear the index
+    pub fn clear_gift_code_tx_out_index(&mut self) {
+        self.gift_code_tx_out_global_index = None;
+    }
+    /// Enable change memos
+    pub fn enable_gift_code_change_memo(&mut self) {
+        self.gift_code_change_memo_enabled = true;
+    }
+    /// Disable change memos
+    pub fn disable_gift_code_change_memo(&mut self) {
+        self.gift_code_change_memo_enabled = false;
+    }
+}
+
+impl MemoBuilder for GiftCodeCancellationMemoBuilder {
+    /// Set the fee
+    fn set_fee(&mut self, _fee: Amount) -> Result<(), NewMemoError> {
+        Ok(())
+    }
+
+    /// Gift code cancellation memos write blank memos to destination TxOut(s)
+    fn make_memo_for_output(
+        &mut self,
+        _amount: Amount,
+        _recipient: &PublicAddress,
+        _memo_context: MemoContext,
+    ) -> Result<MemoPayload, NewMemoError> {
+        Ok(UnusedMemo {}.into())
+    }
+
+    /// Build a memo for a gift code change output
+    fn make_memo_for_change_output(
+        &mut self,
+        _amount: Amount,
+        _change_destination: &ReservedDestination,
+        _memo_context: MemoContext,
+    ) -> Result<MemoPayload, NewMemoError> {
+        if !self.gift_code_change_memo_enabled {
+            return Ok(UnusedMemo {}.into());
+        }
+        if self.wrote_change_memo {
+            return Err(NewMemoError::MultipleChangeOutputs);
+        }
+        if self.gift_code_tx_out_global_index.is_none() {
+            return Err(NewMemoError::MissingInput(
+                "Must specify gift code TxOut global index".into(),
+            ));
+        }
+        self.wrote_change_memo = true;
+        if let Some(tx_out_global_index) = self.gift_code_tx_out_global_index.take() {
+            Ok(GiftCodeCancellationMemo::from(tx_out_global_index).into())
+        } else {
+            Err(NewMemoError::MissingInput(
+                "Missing global index of TxOut to be cancelled".into(),
+            ))
+        }
+    }
```

Comment:
Optional/DRY: Consolidate the MissingInput check
```suggestion
        if let Some(tx_out_global_index) = self.gift_code_tx_out_global_index.take() {
            self.wrote_change_memo = true;
            Ok(GiftCodeCancellationMemo::from(tx_out_global_index).into())
        } else {
            Err(NewMemoError::MissingInput(
                "Missing global index of TxOut to be cancelled".into(),
            ))
        }
    }
```